**SENTIMENT ANALYSIS**  
Performing NLTK VADER sentiment analysis on amazon reviews dataset.

**NLTK's VADER module**  
VADER is an NLTK module that provides sentiment scores based on words used ("completely" boosts a score, while "slightly" reduces it), on capitalization & punctuation ("GREAT!!!" is stronger than "great."), and negations (words like "isn't" and "doesn't" affect the outcome).

**Download the VADER lexicon**. We only need to do this once.

In [1]:
import nltk
nltk.download('vader_lexicon')

[nltk_data] Downloading package vader_lexicon to /root/nltk_data...


True

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


**Import necessary libraries:**

In [3]:
import os
os.chdir("/content/drive/My Drive/Github/")

In [4]:
import numpy as np
import pandas as pd

**Load the Data:**

In [5]:
df = pd.read_csv('amazonreviews.tsv', sep='\t')
df.head()

label                                             review
0   pos  Stuning even for the non-gamer: This sound tra...
1   pos  The best soundtrack ever to anything.: I'm rea...
2   pos  Amazing!: This soundtrack is my favorite music...
3   pos  Excellent Soundtrack: I truly like this soundt...
4   pos  Remember, Pull Your Jaw Off The Floor After He...

**Clean the data:**

In [6]:
# REMOVE NaN VALUES AND EMPTY STRINGS:
df.dropna(inplace=True)

blanks = []  # start with an empty list

for i,lb,rv in df.itertuples():  # iterate over the DataFrame
    if type(rv)==str:            # avoid NaN values
        if rv.isspace():         # test 'review' for whitespace
            blanks.append(i)     # add matching index numbers to the list

df.drop(blanks, inplace=True)

In [7]:
df['label'].value_counts()

neg    5097
pos    4903
Name: label, dtype: int64

**Import SentimentIntensityAnalyzer and create an sid object:**

VADER's **SentimentIntensityAnalyzer()** takes in a string and returns a dictionary of scores in each of four categories:

negative
neutral
positive
compound (computed by normalizing the scores above)

In [8]:
from nltk.sentiment.vader import SentimentIntensityAnalyzer

sid = SentimentIntensityAnalyzer()

/usr/local/lib/python3.7/dist-packages/nltk/twitter/__init__.py:20: UserWarning: The twython library has not been installed. Some functionality from the twitter package will not be available.
  warnings.warn("The twython library has not been installed. "


**Adding Scores and Labels to the DataFrame:**  
Use sid to append a comp_score to the dataset.

In [9]:
df['scores'] = df['review'].apply(lambda review: sid.polarity_scores(review))

df['compound']  = df['scores'].apply(lambda score_dict: score_dict['compound'])

df['comp_score'] = df['compound'].apply(lambda c: 'pos' if c >=0 else 'neg')

df.head()

label                                             review  ... compound  comp_score
0   pos  Stuning even for the non-gamer: This sound tra...  ...   0.9454         pos
1   pos  The best soundtrack ever to anything.: I'm rea...  ...   0.8957         pos
2   pos  Amazing!: This soundtrack is my favorite music...  ...   0.9858         pos
3   pos  Excellent Soundtrack: I truly like this soundt...  ...   0.9814         pos
4   pos  Remember, Pull Your Jaw Off The Floor After He...  ...   0.9781         pos

[5 rows x 5 columns]

**Report on Accuracy:**  
Perform a comparison analysis between the original label and comp_score.

In [10]:
from sklearn.metrics import accuracy_score,classification_report,confusion_matrix

In [11]:
accuracy_score(df['label'],df['comp_score'])

0.7091

In [12]:
print(classification_report(df['label'],df['comp_score']))

              precision    recall  f1-score   support

         neg       0.86      0.51      0.64      5097
         pos       0.64      0.91      0.75      4903

    accuracy                           0.71     10000
   macro avg       0.75      0.71      0.70     10000
weighted avg       0.75      0.71      0.70     10000



In [13]:
print(confusion_matrix(df['label'],df['comp_score']))

[[2623 2474]
 [ 435 4468]]


This tells us that VADER correctly identified an Amazon review as "positive" or "negative" roughly 71% of the time.